File: forgerock-auth/src/test/java/org/forgerock/android/auth/RootDetectorTest.java
/*
 * Copyright (c) 2019 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth;

import android.content.Context;
import androidx.test.core.app.ApplicationProvider;
import org.forgerock.android.auth.detector.*;
import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class RootDetectorTest {

    private Context context = ApplicationProvider.getApplicationContext();

    @Test
    public void testBuildTagDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new BuildTagsDetector())
                .build();
        Assert.assertEquals(1.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testDangerousPropertyDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new DangerousPropertyDetector())
                .build();
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testNativeDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new NativeDetector())
                .build();
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testPermissionDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new PermissionDetector())
                .build();
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testRootApkDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new RootApkDetector())
                .build();
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testRootAppDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new RootAppDetector())
                .build();
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testRootCloakingAppDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new RootCloakingAppDetector())
                .build();
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testRootProgramFileDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new RootProgramFileDetector())
                .build();
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testRootRequiredAppDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new RootRequiredAppDetector())
                .build();
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testSuCommandDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new SuCommandDetector())
                .build();
        Assert.assertEquals(1.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testBusyBoxProgramFileDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new BusyBoxProgramFileDetector())
                .build();
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001);
    }

    @Test
    public void testPartialDetector() {
        RootDetector rootDetector = FRRootDetector.builder()
                .detector(new DangerousPropertyDetector())
                .detector(new NativeDetector())
                .detector(new PermissionDetector())
                .detector(new RootApkDetector())
                .detector(new RootAppDetector())
                .detector(new RootCloakingAppDetector())
                .detector(new RootProgramFileDetector())
                .detector(new RootRequiredAppDetector())
                .detector(new BusyBoxProgramFileDetector())
                .build();
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001);
    }
}

File: forgerock-auth/src/test/java/org/forgerock/android/auth/RootDetectorTest.kt
/*
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 *  This software may be modified and distributed under the terms
 *  of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth

import android.content.Context
import org.robolectric.RobolectricTestRunner
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.forgerock.android.auth.detector.RootDetector
import org.forgerock.android.auth.detector.FRRootDetector
import org.forgerock.android.auth.detector.BuildTagsDetector
import org.forgerock.android.auth.detector.DangerousPropertyDetector
import org.forgerock.android.auth.detector.NativeDetector
import org.forgerock.android.auth.detector.PermissionDetector
import org.forgerock.android.auth.detector.RootApkDetector
import org.forgerock.android.auth.detector.RootAppDetector
import org.forgerock.android.auth.detector.RootCloakingAppDetector
import org.forgerock.android.auth.detector.RootProgramFileDetector
import org.forgerock.android.auth.detector.RootRequiredAppDetector
import org.forgerock.android.auth.detector.SuCommandDetector
import org.forgerock.android.auth.detector.BusyBoxProgramFileDetector
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class RootDetectorTest {
    private val context = ApplicationProvider.getApplicationContext<Context>()
    @Test
    fun testBuildTagDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(BuildTagsDetector())
            .build()
        Assert.assertEquals(1.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testDangerousPropertyDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(DangerousPropertyDetector())
            .build()
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testNativeDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(NativeDetector())
            .build()
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testPermissionDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(PermissionDetector())
            .build()
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testRootApkDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(RootApkDetector())
            .build()
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testRootAppDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(RootAppDetector())
            .build()
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testRootCloakingAppDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(RootCloakingAppDetector())
            .build()
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testRootProgramFileDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(RootProgramFileDetector())
            .build()
        Assert.assertEquals(1.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testRootRequiredAppDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(RootRequiredAppDetector())
            .build()
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testSuCommandDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(SuCommandDetector())
            .build()
        Assert.assertEquals(1.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testBusyBoxProgramFileDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(BusyBoxProgramFileDetector())
            .build()
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001)
    }

    @Test
    fun testPartialDetector() {
        val rootDetector: RootDetector = FRRootDetector.builder()
            .detector(DangerousPropertyDetector())
            .detector(NativeDetector())
            .detector(PermissionDetector())
            .detector(RootApkDetector())
            .detector(RootAppDetector())
            .detector(RootCloakingAppDetector())
            .detector(RootRequiredAppDetector())
            .detector(BusyBoxProgramFileDetector())
            .build()
        Assert.assertEquals(0.0, rootDetector.isRooted(context), 0.0001)
    }
}